src/Buffer.mo

```diff
@@ -1,166 +1,1754 @@
+    /// The predicate is given both the index of the element and the element itself.
+    /// This may cause a downsizing of the array.
+    ///
+    /// Runtime: O(size)
+    ///
+    /// Amortized Space: O(1), Worst Case Space: O(size)
+    public func filterEntries(predicate : (Nat, X) -> Bool) { 
+      var removeCount = 0;
+      let keep = 
+        Prim.Array_tabulate<Bool>(
+          count,
+          func i {
+            switch (elements[i]) {
+              case (?element) {
+                if (predicate(i, element)) {
+                  true
+                } else {
+                  removeCount += 1;
+                  false
+                }
+              };
+              case null {
+                Prim.trap "Malformed buffer in filter()"
+              }
+            }
+          }
+        );
+      
+      let elementsSize = elements.size();
 
-    /// Resets the buffer.
-    public func clear() =
-      count := 0;
+      var original = elements;
+      if ((count - removeCount : Nat) < elementsSize / DOWNSIZE_THRESHOLD) {
+        elements := Prim.Array_init<?X>(elementsSize / DOWNSIZE_FACTOR, null);
+      };
 
-    /// Returns a copy of this buffer.
-    public func clone() : Buffer<X> {
-      let c = Buffer<X>(elems.size());
       var i = 0;
-      label l loop {
-        if (i >= count) break l;
-        c.add(elems[i]);
+      var j = 0;
+      while (i < count) {
+        if (keep[i]) {
+          elements[j] := original[i];
+          i += 1;
+          j += 1;
+        } else {
+          i += 1;
+        }
+      };
+
+      count -= removeCount;
+    };
+
+    /// Gets the `i`-th element of this buffer. Traps if  `i >= count`. Indexing is zero-based.
+    ///
+    /// Runtime: O(1)
+    ///
+    /// Space: O(1)
+    public func get(i : Nat) : X {
+      switch (elements[i]) {
+        case (?element) element;
+        case null Prim.trap("Buffer index out of bounds in get");
+      }
+    };
+
+    /// Gets the `i`-th element of the buffer as an option. Returns `null` when `i >= count`. Indexing is zero-based.
+    ///
+    /// Runtime: O(1)
+    ///
+    /// Space: O(1)
+    public func getOpt(i : Nat) : ?X {
+      if (i < count) {
+        elements[i]
+      } else {
+        null
+      }
+    };
+
+    /// Overwrites the current value of the `i`-entry of  this buffer with `elem`. Traps if the
+    /// index is out of bounds. Indexing is zero-based.
+    ///
+    /// Runtime: O(1)
+    ///
+    /// Space: O(1)
+    public func put(i : Nat, elem : X) {
+      if (i >= count) {
+        Prim.trap "Buffer index out of bounds in put";
+      };
+      elements[i] := ?elem;
+    };
+
+    /// Returns the size of the underlying array.
+    ///
+    /// Runtime: O(1)
+    ///
+    /// Space: O(1)
+    public func capacity() : Nat = elements.size();
+
+    /// Resizes the underyling array to `size`. Traps if `size` is less than
+    /// the current size of the buffer.
```

Comment:
Ah yes that's a good point.

So it actually doesn't resize to `count` necessarily (that is `trim()`), just any value that is at least `count`.

I think I'll change the parameter name to `capacity` (and maybe rename the function), since this function changes the capacity of the buffer (not the size of the logical list as you say).

And I'll also clarify this terminology at the top of the documentation, and rename the constants to follow this nomenclature.

Reply:
I'll only use "size" (and size related compound words, like downsize) to refer to the number of elements in the logical list that the buffer represents.

And I'll use "capacity" to refer to the length of the underlying array.